pkg/pillar/cmd/vaultmgr/vaultmgr.go

```diff
@@ -675,3 +750,76 @@ func handleGlobalConfigDelete(ctxArg interface{}, key string,
 		debugOverride, logger)
 	log.Infof("handleGlobalConfigDelete done for %s\n", key)
 }
+
+// Handles both create and modify events
+func handleVaultKeyFromControllerModify(ctxArg interface{}, key string,
+	keyArg interface{}) {
+	ctx := ctxArg.(*vaultMgrContext)
+	keyFromController, ok := keyArg.(types.EncryptedVaultKeyFromController)
+	if !ok {
+		log.Fatalf("[VAULT] Unexpected pub type %T", keyArg)
+	}
+	if keyFromController.Name != types.DefaultVaultName {
+		log.Warnf("Ignoring unknown vault %s", keyFromController.Name)
+		return
+	}
+	log.Debugf("Processing EncryptedVaultKeyFromController %s\n", key)
+	keyData := &attest.AttestVolumeKeyData{}
+	if err := proto.Unmarshal(keyFromController.EncryptedVaultKey, keyData); err != nil {
+		log.Errorf("Failed to unmarshal keyData %v", err)
+		return
+	}
+	unwrappedKey, err := etpm.UnwrapSecretWithDeviceKey(keyData.EncryptedKey,
+		keyData.PublicKey)
+	if err != nil {
+		log.Errorf("Failed to unwrap keyData %v", err)
+		return
+	}
+	err = sealKeyinTpm(unwrappedKey)
+	if err != nil {
+		log.Errorf("Failed to Seal key in TPM %v", err)
+		return
+	}
+
+	//Try unlocking the vault now, in case it is not yet unlocked
+	if !ctx.defaultVaultUnlocked {
+		err = unlockVault(types.DefaultVaultName, false)
+		if err != nil {
+			log.Errorf("Failed to unlock vault after receiving Controller key, %v",
+				err)
+			return
+		}
+		//fetch, set and publish the latest status
+		publishVaultStatus(ctx)
+	}
+}
+
+func publishVaultKey(ctx *vaultMgrContext, vaultName string) error {
+	keyBytes, err := retrieveTpmKey()
+	if err != nil {
+		return fmt.Errorf("Failed to retrieve key from TPM %v", err)
+	}
+
+	wrappedKey, pubPoint, err := etpm.WrapSecretWithDeviceKey(keyBytes)
+	if err != nil {
+		return fmt.Errorf("Failed to wrap vault key %v", err)
+	}
+
+	keyData := &attest.AttestVolumeKeyData{
+		EncryptedKey: wrappedKey,
+		PublicKey:    pubPoint,
+	}
+	b, err := proto.Marshal(keyData)
+	if err != nil {
+		return fmt.Errorf("Failed to Marshal keyData %v", err)
+	}
+
+	keyFromDevice := types.EncryptedVaultKeyFromDevice{}
+	keyFromDevice.Name = vaultName
+	keyFromDevice.EncryptedVaultKey = b
```

Comment:
Should some comment somewhere say that EncryptedVaultKey is a protobuf encoded message? Or is that a local matter to the vaultmgr?

Reply:
It is an internal detail to vaultmgr IMO

Reply:
ok